File: sdk/python/vouchstone_sdk/graph.py
"""EntityGraph / PolicyGraph / WorkflowTrace — the three-compartment pattern.

Every accountable agent use case (AP-invoice matching, compliance evidence,
data migration, ...) decomposes into the same three compartments:

1. **EntityGraph**  — domain entities and the edges between them (what the
   agent knows: invoices, vendors, POs; controls, evidence, frameworks;
   tables, columns, pipelines; ...). Built from ``Entity`` (see types.py).
2. **PolicyGraph**   — a stable ruleset the agent's proposed actions are
   checked against (permit/forbid + obligations), independent of any one
   domain. Mirrors the shape of the control plane's ABAC engine
   (``app/services/abac.py``) so a policy decision made locally in the
   harness and one made by the hosted gateway are directly comparable.
3. **WorkflowTrace** — an append-only, hash-chained record of what actually
   happened, so a decision can be replayed and verified without a live
   connection to the control plane. Uses the same canonical-JSON + SHA-256
   chaining scheme as the control plane's ledger
   (``app/services/ledger_signing.py``): ``sha256(prev_hash || canonical_json(entry))``.

Before this module, "Specialize" only meant prompt + tools + scoped_subgraph
(see ``AgentConfig.scoped_subgraph``); a use case that needed real domain
structure, a policy check, or a verifiable trace had to hand-roll it. These
three classes formalize that pattern as an SDK primitive: define entities,
define policies, and every ``run()`` call automatically gets an append-only,
verifiable record — no per-use-case SDK code required.
"""

from __future__ import annotations

import hashlib
import json
import re
from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from .types import Entity

# ============================================================
# Canonical JSON + hash chaining
#
# Identical algorithm to control-plane/backend/app/services/ledger_signing.py
# (canonical_json / compute_entry_hash) so a hash computed locally in the
# harness and one computed by the hosted ledger are byte-for-byte
# comparable -- this is what makes a WorkflowTrace independently verifiable
# rather than "trust the SDK's math."
# ============================================================

def canonical_json(payload: dict[str, Any]) -> bytes:
    """RFC 8785-ish canonical JSON: sorted keys, no whitespace, UTF-8."""
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=False, default=str).encode("utf-8")


def compute_entry_hash(prev_hash: str, payload: dict[str, Any]) -> str:
    """sha256(prev_hash || canonical_json(payload))."""
    h = hashlib.sha256()
    h.update((prev_hash or "").encode("utf-8"))
    h.update(canonical_json(payload))
    return h.hexdigest()


# ============================================================
# EntityGraph — domain entities + edges
# ============================================================

@dataclass
class Edge:
    """A directed, typed relationship between two entities."""
    source_id: str
    target_id: str
    edge_type: str
    attributes: dict[str, Any] = field(default_factory=dict)


class EntityGraph:
    """In-memory domain entity graph: nodes are ``Entity`` records, edges are
    typed relationships between them. Not a replacement for the control
    plane's Customer Knowledge Graph -- this is the client-side working set
    an agent reasons over during a single run, typically seeded from
    ``MemoryContext.semantic_entities`` and grown as the agent extracts new
    facts.
    """

    def __init__(self) -> None:
        self._entities: dict[str, Entity] = {}
        self._edges: list[Edge] = []

    def add_entity(self, entity: Entity) -> Entity:
        self._entities[entity.id] = entity
        return entity

    def get_entity(self, entity_id: str) -> Entity | None:
        return self._entities.get(entity_id)

    def entities_by_type(self, entity_type: str) -> list[Entity]:
        return [e for e in self._entities.values() if e.entity_type == entity_type]

    def add_edge(
        self, source_id: str, target_id: str, edge_type: str,
        attributes: dict[str, Any] | None = None,
    ) -> Edge:
        if source_id not in self._entities:
            raise ValueError(f"unknown source entity: {source_id}")
        if target_id not in self._entities:
            raise ValueError(f"unknown target entity: {target_id}")
        edge = Edge(source_id=source_id, target_id=target_id, edge_type=edge_type,
                     attributes=attributes or {})
        self._edges.append(edge)
        return edge

    def related(self, entity_id: str, edge_type: str | None = None) -> list[Entity]:
        """Entities reachable from ``entity_id`` via one outbound edge,
        optionally filtered by edge type."""
        out: list[Entity] = []
        for e in self._edges:
            if e.source_id != entity_id:
                continue
            if edge_type is not None and e.edge_type != edge_type:
                continue
            target = self._entities.get(e.target_id)
            if target is not None:
                out.append(target)
        return out

    def __len__(self) -> int:
        return len(self._entities)

    def all_entities(self) -> list[Entity]:
        return list(self._entities.values())

    def to_dict(self) -> dict[str, Any]:
        return {
            "entities": [
                {
                    "id": e.id, "entity_type": e.entity_type, "entity_key": e.entity_key,
                    "attributes": e.attributes, "confidence": e.confidence,
                    "source_trace_id": e.source_trace_id,
                    "created_at": e.created_at.isoformat() if e.created_at else None,
                }
                for e in self._entities.values()
            ],
            "edges": [
                {"source_id": ed.source_id, "target_id": ed.target_id,
                 "edge_type": ed.edge_type, "attributes": ed.attributes}
                for ed in self._edges
            ],
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> EntityGraph:
        graph = cls()
        for ed in data.get("entities", []):
            created_at = ed.get("created_at")
            graph.add_entity(Entity(
                id=ed["id"], entity_type=ed["entity_type"], entity_key=ed["entity_key"],
                attributes=ed.get("attributes", {}), confidence=ed.get("confidence", 1.0),
                source_trace_id=ed.get("source_trace_id"),
                created_at=datetime.fromisoformat(created_at) if created_at else datetime.now(timezone.utc),
            ))
        for edd in data.get("edges", []):
            graph.add_edge(edd["source_id"], edd["target_id"], edd["edge_type"],
                            edd.get("attributes"))
        return graph


# ============================================================
# PolicyGraph — permit/forbid ruleset with obligations
#
# Same effect/priority/condition shape as app/services/abac.py's evaluator
# so a policy authored once (e.g. in the control plane) can be exported and
# evaluated identically here, offline, by the compatibility gate (C7).
# ============================================================

_CONDITION_OPS: dict[str, Callable[[Any, Any], bool]] = {
    "eq": lambda v, target: v == target,
    "ne": lambda v, target: v != target,
    "in": lambda v, target: v in (target or []),
    "not_in": lambda v, target: v not in (target or []),
    "gt": lambda v, target: v is not None and v > target,
    "gte": lambda v, target: v is not None and v >= target,
    "lt": lambda v, target: v is not None and v < target,
    "lte": lambda v, target: v is not None and v <= target,
    "startswith": lambda v, target: isinstance(v, str) and v.startswith(target),
    "regex": lambda v, target: isinstance(v, str) and re.search(target, v) is not None,
}


def _get_path(root: dict[str, Any], path: str) -> Any:
    cur: Any = root
    for part in path.split("."):
        if isinstance(cur, dict):
            cur = cur.get(part)
        else:
            return None
    return cur


@dataclass
class Policy:
    """One permit/forbid rule. ``priority`` breaks ties (lower runs first);
    the first matching ``forbid`` wins outright, otherwise obligations union
    across every matching ``permit``."""
    name: str
    effect: str  # "permit" | "forbid"
    action: dict[str, Any] | None = None  # e.g. {"eq": "schema.alter_table"}
    resource: dict[str, Any] | None = None  # e.g. {"data_classification": "pii"}
    conditions: list[dict[str, Any]] = field(default_factory=list)  # [{"path", "op", "value"}]
    obligations: list[str] = field(default_factory=list)
    priority: int = 100


@dataclass
class PolicyDecision:
    allow: bool
    obligations: list[str] = field(default_factory=list)
    matched_policy_names: list[str] = field(default_factory=list)
    reason: str = ""


def _matches_action(spec: dict[str, Any] | None, action: str) -> bool:
    if not spec:
        return True
    if "eq" in spec and action != spec["eq"]:
        return False
    if "in" in spec and action not in spec["in"]:
        return False
    if "startswith" in spec and not action.startswith(spec["startswith"]):
        return False
    return True


def _matches_resource(spec: dict[str, Any] | None, resource: dict[str, Any]) -> bool:
    if not spec:
        return True
    for key, target in spec.items():
        value = resource.get(key)
        if isinstance(target, list):
            if value not in target:
                return False
        else:
            if value != target:
                return False
    return True


def _matches_conditions(conditions: list[dict[str, Any]], root: dict[str, Any]) -> bool:
    for cond in conditions:
        path, op, target = cond.get("path"), cond.get("op"), cond.get("value")
        if not path or op not in _CONDITION_OPS:
            return False
        if not _CONDITION_OPS[op](_get_path(root, path), target):
            return False
    return True


class PolicyGraph:
    """A stable, evaluable ruleset. Add policies once; evaluate as many
    proposed actions against them as needed. Deny-by-default: an action
    with no matching permit policy is denied, matching the control plane's
    ABAC posture (see app/services/abac.py's "No matching permit policy
    (default deny)")."""

    def __init__(self) -> None:
        self._policies: list[Policy] = []

    def add_policy(self, policy: Policy) -> Policy:
        self._policies.append(policy)
        return policy

    def evaluate(
        self, *, principal: dict[str, Any], action: str,
        resource: dict[str, Any] | None = None, context: dict[str, Any] | None = None,
    ) -> PolicyDecision:
        resource = resource or {}
        context = context or {}
        root = {"principal": principal, "action": action, "resource": resource, "context": context}

        obligations: list[str] = []
        matched: list[str] = []
        any_permit = False

        for policy in sorted(self._policies, key=lambda p: (p.priority, p.name)):
            if not _matches_action(policy.action, action):
                continue
            if not _matches_resource(policy.resource, resource):
                continue
            if not _matches_conditions(policy.conditions, root):
                continue

            matched.append(policy.name)
            if policy.effect == "forbid":
                return PolicyDecision(
                    allow=False, obligations=[], matched_policy_names=matched,
                    reason=f"forbidden by policy: {policy.name}",
                )
            any_permit = True
            obligations.extend(o for o in policy.obligations if o not in obligations)

        if not any_permit:
            return PolicyDecision(
                allow=False, obligations=[], matched_policy_names=matched,
                reason="no matching permit policy (default deny)",
            )
        return PolicyDecision(allow=True, obligations=obligations, matched_policy_names=matched, reason="allow")


# ============================================================
# WorkflowTrace — append-only, hash-chained record
# ============================================================

@dataclass
class WorkflowTraceEntry:
    sequence: int
    kind: str
    actor: str
    payload: dict[str, Any]
    prev_hash: str
    entry_hash: str
    timestamp: str


class WorkflowTrace:
    """Append-only, hash-chained trace of what an agent (or the
    compatibility gate) actually did. Each entry's hash covers the previous
    entry's hash plus its own canonical payload, so the chain can be
    verified independently -- tampering with or reordering any entry breaks
    every hash after it. This is the client-side counterpart to the control
    plane's signed ledger (``app/services/ledger_signing.py``); a bundle
    pulled offline (C4) carries its own trace that can be verified without
    calling home, and reconciles with the hosted ledger on next sync.
    """

    def __init__(self) -> None:
        self._entries: list[WorkflowTraceEntry] = []

    def append(self, kind: str, payload: dict[str, Any], actor: str = "agent") -> WorkflowTraceEntry:
        prev_hash = self._entries[-1].entry_hash if self._entries else ""
        sequence = len(self._entries) + 1
        timestamp = datetime.now(timezone.utc).isoformat()
        body = {
            "sequence": sequence, "kind": kind, "actor": actor,
            "payload": payload, "timestamp": timestamp,
        }
        entry_hash = compute_entry_hash(prev_hash, body)
        entry = WorkflowTraceEntry(
            sequence=sequence, kind=kind, actor=actor, payload=payload,
            prev_hash=prev_hash, entry_hash=entry_hash, timestamp=timestamp,
        )
        self._entries.append(entry)
        return entry

    @property
    def tip_hash(self) -> str:
        return self._entries[-1].entry_hash if self._entries else ""

    @property
    def entries(self) -> list[WorkflowTraceEntry]:
        return list(self._entries)

    def verify_chain(self) -> bool:
        """Recompute every hash from scratch and confirm the chain is intact."""
        prev_hash = ""
        for entry in self._entries:
            body = {
                "sequence": entry.sequence, "kind": entry.kind, "actor": entry.actor,
                "payload": entry.payload, "timestamp": entry.timestamp,
            }
            if entry.prev_hash != prev_hash:
                return False
            if compute_entry_hash(prev_hash, body) != entry.entry_hash:
                return False
            prev_hash = entry.entry_hash
        return True

    def to_dict(self) -> dict[str, Any]:
        return {
            "tip_hash": self.tip_hash,
            "entries": [
                {
                    "sequence": e.sequence, "kind": e.kind, "actor": e.actor,
                    "payload": e.payload, "prev_hash": e.prev_hash,
                    "entry_hash": e.entry_hash, "timestamp": e.timestamp,
                }
                for e in self._entries
            ],
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> WorkflowTrace:
        trace = cls()
        trace._entries = [
            WorkflowTraceEntry(
                sequence=e["sequence"], kind=e["kind"], actor=e["actor"],
                payload=e["payload"], prev_hash=e["prev_hash"],
                entry_hash=e["entry_hash"], timestamp=e["timestamp"],
            )
            for e in data.get("entries", [])
        ]
        return trace

File: sdk/typescript/src/graph.ts
/**
 * EntityGraph / PolicyGraph / WorkflowTrace — the three-compartment pattern.
 *
 * Mirrors data-plane/sdk/python/vouchstone_sdk/graph.py exactly (same
 * hashing algorithm, same policy semantics) so a bundle produced by one
 * SDK is verifiable by the other. See that file's module docstring for
 * the full rationale.
 */

import { Entity } from './types';

// ============================================================
// Canonical JSON + hash chaining
//
// Identical algorithm to control-plane/backend/app/services/ledger_signing.py
// and the Python SDK's graph.py: sha256(prevHash || canonicalJson(payload)).
// Node's crypto module (not an external dependency) provides SHA-256.
// ============================================================

import { createHash } from 'crypto';

export function canonicalJson(payload: Record<string, any>): string {
  const sortKeys = (value: any): any => {
    if (Array.isArray(value)) return value.map(sortKeys);
    if (value !== null && typeof value === 'object') {
      const sorted: Record<string, any> = {};
      for (const key of Object.keys(value).sort()) {
        sorted[key] = sortKeys(value[key]);
      }
      return sorted;
    }
    return value;
  };
  return JSON.stringify(sortKeys(payload));
}

export function computeEntryHash(prevHash: string, payload: Record<string, any>): string {
  const h = createHash('sha256');
  h.update(prevHash || '', 'utf-8');
  h.update(canonicalJson(payload), 'utf-8');
  return h.digest('hex');
}

// ============================================================
// EntityGraph — domain entities + edges
// ============================================================

export interface GraphEdge {
  sourceId: string;
  targetId: string;
  edgeType: string;
  attributes: Record<string, any>;
}

export class EntityGraph {
  private entitiesById: Map<string, Entity> = new Map();
  private edges: GraphEdge[] = [];

  addEntity(entity: Entity): Entity {
    this.entitiesById.set(entity.id, entity);
    return entity;
  }

  getEntity(entityId: string): Entity | undefined {
    return this.entitiesById.get(entityId);
  }

  entitiesByType(entityType: string): Entity[] {
    return Array.from(this.entitiesById.values()).filter((e) => e.entityType === entityType);
  }

  addEdge(sourceId: string, targetId: string, edgeType: string, attributes: Record<string, any> = {}): GraphEdge {
    if (!this.entitiesById.has(sourceId)) throw new Error(`unknown source entity: ${sourceId}`);
    if (!this.entitiesById.has(targetId)) throw new Error(`unknown target entity: ${targetId}`);
    const edge: GraphEdge = { sourceId, targetId, edgeType, attributes };
    this.edges.push(edge);
    return edge;
  }

  related(entityId: string, edgeType?: string): Entity[] {
    return this.edges
      .filter((e) => e.sourceId === entityId && (edgeType === undefined || e.edgeType === edgeType))
      .map((e) => this.entitiesById.get(e.targetId))
      .filter((e): e is Entity => e !== undefined);
  }

  get size(): number {
    return this.entitiesById.size;
  }

  allEntities(): Entity[] {
    return Array.from(this.entitiesById.values());
  }

  toJSON(): Record<string, any> {
    return {
      entities: Array.from(this.entitiesById.values()),
      edges: this.edges,
    };
  }

  static fromJSON(data: Record<string, any>): EntityGraph {
    const graph = new EntityGraph();
    for (const e of data.entities || []) graph.addEntity(e);
    for (const ed of data.edges || []) graph.addEdge(ed.sourceId, ed.targetId, ed.edgeType, ed.attributes);
    return graph;
  }
}

// ============================================================
// PolicyGraph — permit/forbid ruleset with obligations
// ============================================================

type ConditionOp = 'eq' | 'ne' | 'in' | 'not_in' | 'gt' | 'gte' | 'lt' | 'lte' | 'startswith' | 'regex';

export interface PolicyCondition {
  path: string;
  op: ConditionOp;
  value: any;
}

export interface Policy {
  name: string;
  effect: 'permit' | 'forbid';
  action?: { eq?: string; in?: string[]; startswith?: string };
  resource?: Record<string, any>;
  conditions?: PolicyCondition[];
  obligations?: string[];
  priority?: number;
}

export interface PolicyDecision {
  allow: boolean;
  obligations: string[];
  matchedPolicyNames: string[];
  reason: string;
}

const CONDITION_OPS: Record<ConditionOp, (v: any, target: any) => boolean> = {
  eq: (v, t) => v === t,
  ne: (v, t) => v !== t,
  in: (v, t) => Array.isArray(t) && t.includes(v),
  not_in: (v, t) => !(Array.isArray(t) && t.includes(v)),
  gt: (v, t) => v !== undefined && v !== null && v > t,
  gte: (v, t) => v !== undefined && v !== null && v >= t,
  lt: (v, t) => v !== undefined && v !== null && v < t,
  lte: (v, t) => v !== undefined && v !== null && v <= t,
  startswith: (v, t) => typeof v === 'string' && v.startsWith(t),
  regex: (v, t) => typeof v === 'string' && new RegExp(t).test(v),
};

function getPath(root: Record<string, any>, path: string): any {
  let cur: any = root;
  for (const part of path.split('.')) {
    if (cur !== null && typeof cur === 'object') cur = cur[part];
    else return undefined;
  }
  return cur;
}

function matchesAction(spec: Policy['action'], action: string): boolean {
  if (!spec) return true;
  if (spec.eq !== undefined && action !== spec.eq) return false;
  if (spec.in !== undefined && !spec.in.includes(action)) return false;
  if (spec.startswith !== undefined && !action.startsWith(spec.startswith)) return false;
  return true;
}

function matchesResource(spec: Record<string, any> | undefined, resource: Record<string, any>): boolean {
  if (!spec) return true;
  for (const [key, target] of Object.entries(spec)) {
    const value = resource[key];
    if (Array.isArray(target)) {
      if (!target.includes(value)) return false;
    } else if (value !== target) {
      return false;
    }
  }
  return true;
}

function matchesConditions(conditions: PolicyCondition[] | undefined, root: Record<string, any>): boolean {
  for (const cond of conditions || []) {
    if (!cond.path || !CONDITION_OPS[cond.op]) return false;
    if (!CONDITION_OPS[cond.op](getPath(root, cond.path), cond.value)) return false;
  }
  return true;
}

/**
 * A stable, evaluable ruleset. Deny-by-default: an action with no matching
 * permit policy is denied, matching the control plane's ABAC posture.
 */
export class PolicyGraph {
  private policies: Policy[] = [];

  addPolicy(policy: Policy): Policy {
    this.policies.push({ priority: 100, obligations: [], ...policy });
    return policy;
  }

  evaluate(params: {
    principal: Record<string, any>;
    action: string;
    resource?: Record<string, any>;
    context?: Record<string, any>;
  }): PolicyDecision {
    const resource = params.resource || {};
    const context = params.context || {};
    const root = { principal: params.principal, action: params.action, resource, context };

    const obligations: string[] = [];
    const matched: string[] = [];
    let anyPermit = false;

    const sorted = [...this.policies].sort((a, b) => {
      const pa = a.priority ?? 100;
      const pb = b.priority ?? 100;
      return pa !== pb ? pa - pb : a.name.localeCompare(b.name);
    });

    for (const policy of sorted) {
      if (!matchesAction(policy.action, params.action)) continue;
      if (!matchesResource(policy.resource, resource)) continue;
      if (!matchesConditions(policy.conditions, root)) continue;

      matched.push(policy.name);
      if (policy.effect === 'forbid') {
        return { allow: false, obligations: [], matchedPolicyNames: matched, reason: `forbidden by policy: ${policy.name}` };
      }
      anyPermit = true;
      for (const o of policy.obligations || []) {
        if (!obligations.includes(o)) obligations.push(o);
      }
    }

    if (!anyPermit) {
      return { allow: false, obligations: [], matchedPolicyNames: matched, reason: 'no matching permit policy (default deny)' };
    }
    return { allow: true, obligations, matchedPolicyNames: matched, reason: 'allow' };
  }
}

// ============================================================
// WorkflowTrace — append-only, hash-chained record
// ============================================================

export interface WorkflowTraceEntry {
  sequence: number;
  kind: string;
  actor: string;
  payload: Record<string, any>;
  prevHash: string;
  entryHash: string;
  timestamp: string;
}

export class WorkflowTrace {
  private entryList: WorkflowTraceEntry[] = [];

  append(kind: string, payload: Record<string, any>, actor: string = 'agent'): WorkflowTraceEntry {
    const prevHash = this.entryList.length > 0 ? this.entryList[this.entryList.length - 1].entryHash : '';
    const sequence = this.entryList.length + 1;
    const timestamp = new Date().toISOString();
    const body = { sequence, kind, actor, payload, timestamp };
    const entryHash = computeEntryHash(prevHash, body);
    const entry: WorkflowTraceEntry = { sequence, kind, actor, payload, prevHash, entryHash, timestamp };
    this.entryList.push(entry);
    return entry;
  }

  get tipHash(): string {
    return this.entryList.length > 0 ? this.entryList[this.entryList.length - 1].entryHash : '';
  }

  get entries(): WorkflowTraceEntry[] {
    return [...this.entryList];
  }

  verifyChain(): boolean {
    let prevHash = '';
    for (const entry of this.entryList) {
      const body = { sequence: entry.sequence, kind: entry.kind, actor: entry.actor, payload: entry.payload, timestamp: entry.timestamp };
      if (entry.prevHash !== prevHash) return false;
      if (computeEntryHash(prevHash, body) !== entry.entryHash) return false;
      prevHash = entry.entryHash;
    }
    return true;
  }

  toJSON(): Record<string, any> {
    return { tipHash: this.tipHash, entries: this.entryList };
  }

  static fromJSON(data: Record<string, any>): WorkflowTrace {
    const trace = new WorkflowTrace();
    trace.entryList = data.entries || [];
    return trace;
  }
}
